doc/watermarks_HLD.md

```diff
@@ -0,0 +1,352 @@
+# 2 Requirements
+The following watermarks should be supported:
+## 2.1 Watermark counters requirements
+|                                          | SAI attribute mapping  |
+|------------------------------------------|------------------------|
+| Ingress headroom per PG                  | SAI_INGRESS_PRIORITY_GROUP_STAT_XOFF_ROOM_WATERMARK_BYTES            |
+| Ingress shared pool occupancy per PG                | SAI_INGRESS_PRIORITY_GROUP_STAT_SHARED_WATERMARK_BYTES            |
+| Egress shared pool occupancy per queue (including both unicast queues and multicast queues)      | SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES         |
+
+System behavior:
+We consider a maximum of one regular user and a maximum of one special user that comes from streaming telemetry (grpc)
+
+Streaming telemetry is only interested in periodic watermark, i.e., it queries the watermark at regular intervals. The interval is configurable. Streaming telemetry does not care about persistent watermark. 
+
+Regular user is able to query the watermark over a period specified in the cli. When the cli is issued, the cli blocks over the specified period, which counts the issuing of the cli as its starting point, and then returns the watermark over the specified period.
+Regular user is able to query the persistent watermark. Regular user is able to reset the persistent watermark. When the persistent watermark is reset, persistent watermark starts a new recording from the time reset is issued.
+
+When one regular user and the streaming telemetry coexist, they do not interfere with each other. Their behaviors stay the same as described above. So the software should be able to handle the following situations and return the correct watermark values to each user:
+
+![](https://github.com/mykolaf/SONiC/blob/gh-pages/images/watermark_HLD/timeline.png)
+
+# 3 Modules Design
+## 3.1 Modules that need to be updated
+
+### 3.1.1 Counter DB
+
+#### The following new Queue counters should be available for each queue entry in the DB:
+- "COUNTERS:queue_vid"
+  - SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES
+#### For every Ingress PG the following should be available in the DB:
+- "COUNTERS:pg_vid"
+  - SAI_INGRESS_PRIORITY_GROUP_STAT_XOFF_ROOM_WATERMARK_BYTES
+  - SAI_INGRESS_PRIORITY_GROUP_STAT_SHARED_WATERMARK_BYTES
+#### Additionally a few mappings should be added:
+- "COUNTERS_PG_PORT_MAP" - map PG oid to port oid
+- "COUNTERS_PG_NAME_MAP" - map PG oid to PG name
+- "COUNTERS_PG_INDEX_MAP" - map PG oid to PG index
+
+The watermark counters are provided via Flex Counter, with a period of 1s. Flex Counter does clear the value from HW. 
+
+#### New tables will be introduced: 
+
+| Table  | Updated by | Cleared by | Used by | Purpose |
+| ------------- | ------------- | --- | --- | --- |
+| COUNTERS  | Flex counter  | No need to clear, Flex Counter clears the value on HW every 1s(by default) and overwrites the DB | Lua plugins(Flex counter plugins) | Contains the counters updated by Flex counters |
```

Comment:
can you add the schema you want to introduce for read-on-clear for a group of counters?